orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -1202,8 +1203,91 @@ def load_state_dict(self, state_dict, strict=True):
             return
 
         # load states onto the frontend onnx graph
-        state_dict = self._load_state_dict_impl(state_dict, strict=strict)
+        optimizer_state_dict = self._load_state_dict_impl(state_dict, strict=strict)
 
         # create a new training session after loading initializer states onto the onnx graph
         # pass the populated states to the training session to populate the backend graph
-        self._init_session(state_dict)
+        self._init_session(optimizer_state_dict)
+
+    def save_checkpoint(self, path, user_dict={}, pytorch_format=False):
+        """Persists ORTTrainer state dictionary on disk along with user_dict.
```

Comment:
This is saved as an hdf5 file format. So, users may not be able to load the file from outside unless they know that the file format is an hdf5 implementation. Any file saved with ORTTrainer.save_checkpoint must be only loaded with ORTTrainer.load_checkpoint.
Having said that, I agree that the name is a little misleading. If a file is saved in the pytorch_format (only model states), there is no way it can be aggregated correctly (since we don;t have the partition_info and training_options subdicts any more). So, the structure has to remain like this while saving (hierarchical).
We could change the name to ```include_optimizer_states``` and I think that makes sense here. I will make the necessary changes for this.

Reply:
Good point. 

Considering the original API, if pytorch_format=True, we could save using torch.save and this would really be compatible. If false, we use hdf5?

Considering the new approach, we would have to update the design doc, as it mentions pytorch_format flag there. 

The documentation in the design (including samples) and in the docstring should remove this ambiguity of whether it can or cannot be loaded into pytorch after saving with True

Let's consider the end-to-end use case for saving/loading for pytorch use here and see what is more natural

Reply:
We could try that approach for non zero runs. For ZeRO aggregation is required to move to PyTorch. Currently ,aggregation only happens on loading and not during saving (which is what you're proposing). Aggregation while saving would require a co-ordination among ranks and this might be possible but might require more investigation.
I think this would be a good item for future improvement. But for now, I think we can continue with our current implementation, and make incremental improvements as and when it becomes possible.